File: 209-minimum-size-subarray-sum/209-minimum-size-subarray-sum.cpp
class Solution {
public:
    int minSubArrayLen(int target, vector<int>& nums) {
        int sum=0,ans=INT_MAX;
        for(int i=0,j=0;i<nums.size();i++){
            sum+=nums[i];
            if(sum>=target){
                ans = min(ans,i-j+1);
                while(sum>=target){
                    sum-=nums[j];
                    ans = min(ans,i-j+1);
                    j++;
                }
            }
        }
        
        return ans==INT_MAX?0:ans;
    }
};

File: 209-minimum-size-subarray-sum/209-minimum-size-subarray-sum.java
class Solution {
    public int minSubArrayLen(int target, int[] nums) {
        int l=0,r=nums.length;
        while(l<=r){
            int m = (l+r)/2,sum=0,ans=-1;
            
            for(int i=0;i<nums.length;i++){
                sum+=nums[i];
                if(i>=m)
                    sum-=nums[i-m];
                ans=Math.max(ans,sum);
            }
            
            if(ans>=target){
                r=m-1;
            }else{
                l=m+1;
            }
        }
        return l>nums.length?0:l;
    }
}